src/backend/distributed/utils/shard_utils.c

```diff
@@ -55,3 +59,52 @@ GetLongestShardName(Oid citusTableOid, char *finalRelationName)
 
 	return longestShardName;
 }
+
+
+/*
+ * GetLongestHypotheticalShardName is a utility function that creates a hypothetical shard
+ * name for a table that is not distributed yet.
+ */
+char *
+GetLongestHypotheticalShardName(Oid parentTableOid, char *relationName)
+{
+	char *longestShardName = pstrdup(relationName);
+	CitusTableCacheEntry *cacheEntry = GetCitusTableCacheEntry(parentTableOid);
+	int shardIntervalCount = cacheEntry->shardIntervalArrayLength;
+
+	Oid savedUserId = InvalidOid;
+	int savedSecurityContext = 0;
+
+	text *sequenceName = cstring_to_text(SHARDID_SEQUENCE_NAME);
+	Oid sequenceId = ResolveRelationId(sequenceName, false);
+	Datum sequenceIdDatum = ObjectIdGetDatum(sequenceId);
+
+	GetUserIdAndSecContext(&savedUserId, &savedSecurityContext);
+	SetUserIdAndSecContext(CitusExtensionOwner(), SECURITY_LOCAL_USERID_CHANGE);
+
+	int64 newShardId = 0;
+
+	/*
+	 * pg_sequence_last_value() returns NULL if the sequence value is not yet used.
+	 * DirectFunctionCall1() gives an ERROR message on NULL return values, and that's why we
+	 * need a PG_TRY block.
+	 */
+	PG_TRY();
+	{
+		Datum lastShardIdDatum = DirectFunctionCall1(pg_sequence_last_value,
+													 sequenceIdDatum);
+		newShardId = DatumGetInt64(lastShardIdDatum) + shardIntervalCount;
+	}
+	PG_CATCH();
+	{
+		/* assume that we have a shardId with 7 digits */
+		newShardId = 9999999;
```

Comment:
We should maybe mark `newShardId` as `volatile`, as the comment suggests in `PG_TRY()` definition:
```
 * Note: if a local variable of the function containing PG_TRY is modified
 * in the PG_TRY section and used in the PG_CATCH section, that variable
 * must be declared "volatile" for POSIX compliance.  This is not mere
 * pedantry; we have seen bugs from compilers improperly optimizing code
 * away when such a variable was not marked.  Beware that gcc's -Wclobbered
 * warnings are just about entirely useless for catching such oversights.
```

We are not using `newShardId` but lets be robust to future changes